src/hotspot/share/gc/shenandoah/shenandoahAgeCensus.cpp

```diff
@@ -0,0 +1,271 @@
+    _local_age_table = NEW_C_HEAP_ARRAY(AgeTable*, max_workers, mtGC);
+    CENSUS_NOISE(_local_noise = NEW_C_HEAP_ARRAY(ShenandoahNoiseStats, max_workers, mtGC);)
+    for (uint i = 0; i < max_workers; i++) {
+      _local_age_table[i] = new AgeTable(false);
+      CENSUS_NOISE(_local_noise[i].clear();)
+    }
+  }
+  _epoch = MAX_SNAPSHOTS - 1;  // see update_epoch()
+}
+
+CENSUS_NOISE(void ShenandoahAgeCensus::add(uint obj_age, uint region_age, uint region_youth, size_t size, uint worker_id) {)
+NO_CENSUS_NOISE(void ShenandoahAgeCensus::add(uint obj_age, uint region_age, size_t size, uint worker_id) {)
+  if (obj_age <= markWord::max_age) {
+    assert(obj_age < MAX_COHORTS && region_age < MAX_COHORTS, "Should have been tenured");
+#ifdef SHENANDOAH_CENSUS_NOISE
+    // Region ageing is stochastic and non-monotonic; this vitiates mortality
+    // demographics in ways that might defeat our algorithms. Marking may be a
+    // time when we might be able to correct this, but we currently do not do
+    // this. Like skipped statistics further below, we want to track the
+    // impact of this noise to see if this may be worthwhile. JDK-<TBD>.
+    uint age = obj_age;
+    if (region_age > 0) {
+      add_aged(size, worker_id);   // this tracking is coarse for now
+      age += region_age;
+      if (age >= MAX_COHORTS) {
+        age = (uint)(MAX_COHORTS - 1);  // clamp
+        add_clamped(size, worker_id);
+      }
+    }
+    if (region_youth > 0) {   // track object volume with retrograde age
+      add_young(size, worker_id);
+    }
+#else   // SHENANDOAH_CENSUS_NOISE
+    uint age = MIN2(obj_age + region_age, (uint)(MAX_COHORTS - 1));  // clamp
+#endif  // SHENANDOAH_CENSUS_NOISE
+    get_local_age_table(worker_id)->add(age, size);
+  } else {
+    // update skipped statistics
+    CENSUS_NOISE(add_skipped(size, worker_id);)
+  }
+}
+
+#ifdef SHENANDOAH_CENSUS_NOISE
+void ShenandoahAgeCensus::add_skipped(size_t size, uint worker_id) {
+  _local_noise[worker_id].skipped += size;
+}
+
+void ShenandoahAgeCensus::add_aged(size_t size, uint worker_id) {
+  _local_noise[worker_id].aged += size;
+}
+
+void ShenandoahAgeCensus::add_clamped(size_t size, uint worker_id) {
+  _local_noise[worker_id].clamped += size;
+}
+
+void ShenandoahAgeCensus::add_young(size_t size, uint worker_id) {
+  _local_noise[worker_id].young += size;
+}
+#endif // SHENANDOAH_CENSUS_NOISE
+
+// Update the epoch for the global age tables,
+// and merge local age tables into the global age table.
+// If appropriate, compute the new tenuring threshold for
+// this epoch.
+void ShenandoahAgeCensus::update_epoch() {
+  assert(_epoch < MAX_SNAPSHOTS, "Out of bounds");
+  if (++_epoch >= MAX_SNAPSHOTS) {
+    _epoch=0;
+  }
+  // Merge data from local age tables into the global age table for the epoch,
+  // clearing the local tables.
+  _global_age_table[_epoch]->clear();
+  CENSUS_NOISE(_global_noise[_epoch].clear();)
+  if (!GenShenCensusAtEvac) {
```

Comment:
In the case of `GenShenCensusAtEvac` this work is done by the calling method explicitly via `ingest` and `compute_...` 

This should probably be done in a separate method/API rather to keep the API/spec of this method clean.

Reply:
Done